### PR TITLE
fix: update BucketInfo translation code to properly handle lifecycle rules

### LINKED ISSUE #850: Storage#update always removes all lifecycle rules if they are not specified
Issue with `com.google.cloud.storage.Storage#update(com.google.cloud.storage.BucketInfo, com.google.cloud.storage.Storage.BucketTargetOption...)`

#### Expected behaviour:
update only fields that are specified in `BucketInfo` parameter.

#### Actual behaviour:
along the updated fields, removes lifecycle rules if they are not specified.

Example code:
```scala
    val bucketId = "[...]"

    val client = StorageOptions.getDefaultInstance.getService
    val before = client.get(bucketId)
    println(s"Before: ${before.getLifecycleRules.size()}")

    val updateRequest = BucketInfo
      .newBuilder(bucketId)
      .setLabels(singletonMap("test", "test"))
      .build
    client.update(updateRequest)

    val after = client.get(bucketId)
    println(s"After: ${after.getLifecycleRules.size()}")
```

Actual output (for bucket with 1 lifecycle rule):
```txt
Before: 1
After: 0
```

I suspect that issue is located here:
https://github.com/googleapis/java-storage/blob/accf5021e99a083aa4b3e8935d8ccd0e3668ae6f/google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java#L1841-L1869

### PR DESCRIPTION
**Pull request by BenWhitehead**
Fixes #850

### CODE REVIEW

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java [hunk 1]
```diff
@@ -1862,7 +1862,7 @@ public Rule apply(LifecycleRule lifecycleRule) {
               }));
     }
 
-    if (rules != null) {
+    if (!rules.isEmpty()) {
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:1865]: Will this support rules deletion as well? I'm wondering if rules should be set to null at construction time and only have a non-null value when rules are returned.

Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:1865]: I'll test and update if necessary

Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:1865]: kokoro failed on this case

Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java]: Added some unit tests and refactored things afterward to account for clearing rules as well.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java [hunk 2]
```diff
@@ -1838,33 +1839,44 @@ public ObjectAccessControl apply(Acl acl) {
...
-    }
 
-    if (rules != null) {
+    if (deleteRules != null || lifecycleRules != null) {
       Lifecycle lifecycle = new Lifecycle();
-      lifecycle.setRule(ImmutableList.copyOf(rules));
+
+      if ((deleteRules == null && lifecycleRules.isEmpty())
+          || (lifecycleRules == null && deleteRules.isEmpty())
+          || (deleteRules != null && deleteRules.isEmpty() && lifecycleRules.isEmpty())) {
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:1848]: This got very complex, could you add a comment above this condition to clarify why it's needed?

**Review by frankyn — CHANGES_REQUESTED** [review]

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java (hunk 2 above)
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:1856]: Comment added PTAL

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java [hunk 3]
```diff
@@ -1843,6 +1843,14 @@ public ObjectAccessControl apply(Acl acl) {
...
+      // Here we determine if we need to "clear" any defined Lifecycle rules by explicitly setting
+      // the Rule list of lifecycle to the empty list.
+      // In order for us to clear the rules, one of the three following must be true:
+      //   1. deleteRules is null while lifecycleRules is non-null and empty
+      //   2. lifecycleRules is null while deleteRules is non-null and empty
+      //   3. lifecycleRules is non-null and empty while deleteRules is non-null and empty
+      // If none of the above three is true, we will interpret as the Lifecycle rules being
+      // updated to the defined set of DeleteRule and LifecycleRule.
       if ((deleteRules == null && lifecycleRules.isEmpty())
           || (lifecycleRules == null && deleteRules.isEmpty())
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:1855]: should you check that `deleteRules != null` here as well?

**Review by frankyn — CHANGES_REQUESTED** [review]

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java (hunk 3 above)
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:1855]: `deleteRules != null` is already true by the time we reach this branch of the if statement, and by virtue of the outer guarding if statement. Either `lifecycleRules` or `deleteRules` is non-null, so there is no need to check it again.

**Review by frankyn — APPROVED** [review]

**Pull request merged**